### ISSUE TITLE
Porting code based on uproot3 to uproot4 - interpretations with custom-dtypes

### ISSUE THREAD

**Issue by tamasgal** [scikit-hep/uproot4#76]
I started porting our I/O library based on uproot3 to uproot4 and although I know that the docs are not done yet, I am also hoping to solve some performance issues which are related to `dtypes` falling back to `object`-type, so I thought I give it a try.

One of the current issues is that I have not figured out yet how to pass custom interpretations to `.array()`. This is the uproot3 code:

```python
import uproot

f = uproot.open("km3net_online.root")

dtype = [("dom_id", "i4"), ("dq_status", "u4"), ("hrv", "u4"), ("fifo", "u4"), ("status3", "u4"), ("status4", "u4")] + [(f"ch{c}", "u1") for c in range(31)]
path = "KM3NET_SUMMARYSLICE/KM3NET_SUMMARYSLICE/vector<KM3NETDAQ::JDAQSummaryFrame>"

data = f[path].array(uproot.asjagged(uproot.astable(uproot.asdtype(dtype)), skipbytes=10))                                                              

data.dtype  # dtype('O')

data[0]   # Out[8]: <Table [<Row 0> <Row 1> <Row 2> ... <Row 61> <Row 62> <Row 63>] at 0x7ff5eb6586d0>

data[0].dtype  # dtype([('dom_id', '<i4'), ('dq_status', '<u4'), ('hrv', '<u4'), ('fifo', '<u4'), ('status3', '<u4'), ('status4', '<u4'), ('ch0', 'u1'), ('ch1', 'u1'), ('ch2', 'u1'), ('ch3', 'u1'), ('ch4', 'u1'), ('ch5', 'u1'), ('ch6', 'u1'), ('ch7', 'u1'), ('ch8', 'u1'), ('ch9', 'u1'), ('ch10', 'u1'), ('ch11', 'u1'), ('ch12', 'u1'), ('ch13', 'u1'), ('ch14', 'u1'), ('ch15', 'u1'), ('ch16', 'u1'), ('ch17', 'u1'), ('ch18', 'u1'), ('ch19', 'u1'), ('ch20', 'u1'), ('ch21', 'u1'), ('ch22', 'u1'), ('ch23', 'u1'), ('ch24', 'u1'), ('ch25', 'u1'), ('ch26', 'u1'), ('ch27', 'u1'), ('ch28', 'u1'), ('ch29', 'u1'), ('ch30', 'u1')])

data[0].dom_id

# array([806451572, 806455814, 806465101, 806483369, 806487219, 806487226,
#       806487231, 808432835, 808435278, 808447180, 808447186, 808451904,
#       808451907, 808469129, 808472260, 808472265, 808488895, 808488990,
#    ...
#       808981864, 808982018, 808982041, 808982077, 808982547, 808984711,
#       808996773, 808997793, 809006037, 809007627, 809503416, 809521500,
#       809524432, 809526097, 809544058, 809544061], dtype=int32)
```

In `uproot4`, this is what I came up with

```python
# similar lines as above until the branch interpretation

interp = uproot4.interpretation.jagged.AsJagged(uproot4.interpretation.numerical.AsDtype(dtype), header_bytes=10)
# AsJagged(AsDtype("[('dom_id', '<i4'), ('dq_status', '<u4'), ('hrv', '<u4'), ('fifo', '<u4'), ('status3', '<u4'), ('status4', '<u4'), ('ch0', 'u1'), ('ch1', 'u1'), ('ch2', 'u1'), ('ch3', 'u1'), ('ch4', 'u1'), ('ch5', 'u1'), ('ch6', 'u1'), ('ch7', 'u1'), ('ch8', 'u1'), ('ch9', 'u1'), ('ch10', 'u1'), ('ch11', 'u1'), ('ch12', 'u1'), ('ch13', 'u1'), ('ch14', 'u1'), ('ch15', 'u1'), ('ch16', 'u1'), ('ch17', 'u1'), ('ch18', 'u1'), ('ch19', 'u1'), ('ch20', 'u1'), ('ch21', 'u1'), ('ch22', 'u1'), ('ch23', 'u1'), ('ch24', 'u1'), ('ch25', 'u1'), ('ch26', 'u1'), ('ch27', 'u1'), ('ch28', 'u1'), ('ch29', 'u1'), ('ch30', 'u1')]"), header_bytes=10)

f[path].array(interp)  # <Array [[{dom_id: 1954091312, ... ch30: 48}]] type='3 * var * {"dom_id": int32, ...'>
```

The first problem is that the endianness is not picked up correctly, the `dom_id` for example should be little endian:

```python
import numpy as np

np.array(f[path].array(interp)[0].dom_id).byteswap()[:4]
# array([806451572, 806455814, 806465101, 806483369], dtype=int32)
```
The second problem is that I am not sure if it is now a "type-safe" (high-performance) readout. In case of `uproot3`, the `dtype='O'` really had a horrible performance, for obvious reasons. Now I think that with `awkward1` I should be able to correctly map the structure, or am I wrong? At least what I got here is a `awkward1.highlevel.Array`.

However, checking the `dtype` is giving me an error and it seems that it is an `object` type:

```python
>>> f[path].array(interp).dtype                                                                                                    
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
<ipython-input-52-98270c2c8c48> in <module>
----> 1 f[path].array(interp).dtype

~/.virtualenvs/km3net/lib/python3.8/site-packages/awkward1/highlevel.py in __getattr__(self, where)
   1015         """
   1016         if where in dir(type(self)):
-> 1017             return super(Array, self).__getattribute__(where)
   1018         else:
   1019             if where in self._layout.keys():

~/.virtualenvs/km3net/lib/python3.8/site-packages/awkward1/_connect/_pandas.py in dtype(self)
    180 
    181         else:
--> 182             return np.dtype(np.object)
    183 
    184     @property

AttributeError: 'NumpyMetadata' object has no attribute 'object'
```

Thanks in advance!

Here is the file: 
[km3net_online.root.zip](https://github.com/scikit-hep/uproot4/files/5123202/km3net_online.root.zip)

**Comment by tamasgal** [comment]
I am scratching my head now... I copy-pasted the code above and now I get

```python
>>> f[path].array(interp)
TypeError: branch interpretation must be an Interpretation, not AsJagged(AsDtype("[('dom_id', '<i4'), ('dq_status', '<u4'), ('hrv', '<u4'), ('fifo', '<u4'), ('status3', '<u4'), ('status4', '<u4'), ('ch0', 'u1'), ('ch1', 'u1'), ('ch2', 'u1'), ('ch3', 'u1'), ('ch4', 'u1'), ('ch5', 'u1'), ('ch6', 'u1'), ('ch7', 'u1'), ('ch8', 'u1'), ('ch9', 'u1'), ('ch10', 'u1'), ('ch11', 'u1'), ('ch12', 'u1'), ('ch13', 'u1'), ('ch14', 'u1'), ('ch15', 'u1'), ('ch16', 'u1'), ('ch17', 'u1'), ('ch18', 'u1'), ('ch19', 'u1'), ('ch20', 'u1'), ('ch21', 'u1'), ('ch22', 'u1'), ('ch23', 'u1'), ('ch24', 'u1'), ('ch25', 'u1'), ('ch26', 'u1'), ('ch27', 'u1'), ('ch28', 'u1'), ('ch29', 'u1'), ('ch30', 'u1')]"), header_bytes=10) (type <class 'uproot4.interpretation.jagged.AsJagged'>)
```

This definitely worked before since I also found it in my history, I should have logged the `uproot4` version :see_no_evil:

**Comment by jpivarski** [comment]
The way this works is not fundamentally different from before: the interpretation is still the first argument of the `array` method. I don't know why it should be saying that `AsJagged` is not an `Interpretation`: it definitely is!

```python
>>> interp
AsJagged(AsDtype("[('dom_id', '<i4'), ('dq_status', '<u4'), ('hrv', '<u4'), ('fifo', '<u4'), ('status3', '<u4'), ('status4', '<u4'), ('ch0', 'u1'), ('ch1', 'u1'), ('ch2', 'u1'), ('ch3', 'u1'), ('ch4', 'u1'), ('ch5', 'u1'), ('ch6', 'u1'), ('ch7', 'u1'), ('ch8', 'u1'), ('ch9', 'u1'), ('ch10', 'u1'), ('ch11', 'u1'), ('ch12', 'u1'), ('ch13', 'u1'), ('ch14', 'u1'), ('ch15', 'u1'), ('ch16', 'u1'), ('ch17', 'u1'), ('ch18', 'u1'), ('ch19', 'u1'), ('ch20', 'u1'), ('ch21', 'u1'), ('ch22', 'u1'), ('ch23', 'u1'), ('ch24', 'u1'), ('ch25', 'u1'), ('ch26', 'u1'), ('ch27', 'u1'), ('ch28', 'u1'), ('ch29', 'u1'), ('ch30', 'u1')]"))
>>> isinstance(interp, uproot4.interpretation.Interpretation)
True
```

Endianness works the same way: the endianness of the `from_dtype` is what you specify, but the `to_dtype` is the native endian version of that if `to_dtype` is not specified:

https://github.com/scikit-hep/uproot4/blob/3827f967da9c1031c8b5de14a877184d1089470d/uproot4/interpretation/numerical.py#L129-L135

The arguments involving `header_bytes` might have changed since they were a bit of a mess before and I tried to consolidate that. Updating your use of `header_bytes` options might involve some investigation of the code; sorry.

One of the new things that can help is that TBranch objects now have `debug` and `debug_array` methods built in, which we used to use the `uproot.asdebug` for. I've put a lot of bells and whistles into the debugging routines (centrally located in Cursor) because I had spent so much time working on these reverse engineering problems. You probably want to take a look at

```python
f["KM3NET_SUMMARYSLICE/KM3NET_SUMMARYSLICE/vector<KM3NETDAQ::JDAQSummaryFrame>"].debug(0)
```

Another new thing is "[strided objects](https://github.com/scikit-hep/uproot4/blob/3827f967da9c1031c8b5de14a877184d1089470d/uproot4/interpretation/objects.py#L403)," which were `asobj(astable(...))` before. Strided means that they are interpreted as interleaving strides of a buffer that was loaded all at once, not a Python for loop. The other kind of object interpretation, "[as objects](https://github.com/scikit-hep/uproot4/blob/3827f967da9c1031c8b5de14a877184d1089470d/uproot4/interpretation/objects.py#L113)," is a Python for loop. I think the only difference between a jagged, strided object interpretation and a jagged, recordarray-dtype interpretation is that the strided objects get labeled with a parameter that Awkward can pick up and use to add methods, such as adding boosting methods to Lorentz vectors. The structure of these two cases is the same (and parameters can always be added later).

**Comment by tamasgal** [comment]
Alright thanks! My main confusion came from the fact that in `uproot3` I used `('dom_id', 'i4')` whereas in `uproot4` I had to specify `('dom_id', '>i4')`. Anyways, I had a busy day and should have spent a bit more time on investigating.

The `debug` stuff is really great, I really appreciate that!

The strided objects is an interesting feature, I have to explore it.

Btw. I just stumbled upon another issue while I am trying to evaluate the performance of the new awkward1-arrays returned by `uproot4`. They seem to work really nice and intuitively, I am happy that I can get rid of the user-interface-code which I designed to propagate nested indexing over a hierarchy of numpy recarrays which were mapped to lazyarrays of ROOT locations (long ones, just as you saw before). Now I only need to take care of a user-friendly mapping and people can simply work with awkward1 arrays, very nice!

...so the issue I stumbled upon is a memory-pileup when executing `branch.array(interp)` multiple times on a larger file. I'll create a new issue with an MWE.
